shared/src/pool_cache.rs

```diff
@@ -1,125 +1,176 @@
-use crate::{
-    current_block::{self, CurrentBlockStream},
-    maintenance::Maintaining,
-    pool_fetching::{Block, Pool, PoolFetching},
-};
-use anyhow::{Context, Result};
+//! Module containing a generic cache for recent on-chain data.
+//!
+//! The design of this module is driven by the need to always return data quickly so that end users
+//! going through the api do not have to wait longer than necessary:
+//! - The mutex is never locked while waiting on an async operation (getting on-chain data from the node).
+//! - Automatically updating the cache is decoupled from normal on-chain data fetches.
+//!
+//! A result of this is that it is possible that the same uncached entry is requested multiple times
+//! simultaneously and some work is wasted. This is unlikely to happen in practice and the value is
+//! going to be cached the next time it is needed.
+//!
+//! When entries are requested we mark all those entries as recently used which potentially evicts
+//! other entries from the lru cache. Cache misses are fetched and inserted into the cache.
+//! Then when the automatic update runs the next time, we request and cache all recently used
+//! entries. For some consumers we only care about the "recent" state of the entries. So we can
+//! return any result from the cache even if it comes from previous blocks.
+//!
+//! On the other hand for others we need to fetch on-chain data at exact blocks which is why we keep
```

Comment:
NIT: I think this new paragraph is meant to stay with the last sentence from the previous. Otherwise, I am not sure what "on the other hand" is referring to.